modules/mod_facts.c

```diff
@@ -212,75 +215,70 @@ static size_t facts_mlinfo_fmt(struct mlinfo *info, char *buf, size_t bufsz) {
   ptr = buf;
 
   if (facts_opts & FACTS_OPT_SHOW_MODIFY) {
-    snprintf(ptr, bufsz, "modify=%04d%02d%02d%02d%02d%02d;",
+    len = snprintf(ptr, bufsz, "modify=%04d%02d%02d%02d%02d%02d;",
       info->tm->tm_year+1900, info->tm->tm_mon+1, info->tm->tm_mday,
       info->tm->tm_hour, info->tm->tm_min, info->tm->tm_sec);
-    buflen = strlen(buf);
+    buflen += len;
     ptr = buf + buflen;
   }
 
   if (facts_opts & FACTS_OPT_SHOW_PERM) {
-    snprintf(ptr, bufsz - buflen, "perm=%s;", info->perm);
-    buflen = strlen(buf);
+    len = snprintf(ptr, bufsz - buflen, "perm=%s;", info->perm);
+    buflen += len;
     ptr = buf + buflen;
   }
 
   if (!S_ISDIR(info->st.st_mode) &&
       (facts_opts & FACTS_OPT_SHOW_SIZE)) {
-    snprintf(ptr, bufsz - buflen, "size=%" PR_LU ";",
+    len = snprintf(ptr, bufsz - buflen, "size=%" PR_LU ";",
       (pr_off_t) info->st.st_size);
-    buflen = strlen(buf);
+    buflen += len;
     ptr = buf + buflen;
   }
 
   if (facts_opts & FACTS_OPT_SHOW_TYPE) {
-    snprintf(ptr, bufsz - buflen, "type=%s;", info->type);
-    buflen = strlen(buf);
+    len = snprintf(ptr, bufsz - buflen, "type=%s;", info->type);
+    buflen += len;
     ptr = buf + buflen;
   }
 
   if (facts_opts & FACTS_OPT_SHOW_UNIQUE) {
-    snprintf(ptr, bufsz - buflen, "unique=%lXU%lX;",
+    len = snprintf(ptr, bufsz - buflen, "unique=%lXU%lX;",
       (unsigned long) info->st.st_dev, (unsigned long) info->st.st_ino);
-    buflen = strlen(buf);
+    buflen += len;
     ptr = buf + buflen;
   }
 
   if (facts_opts & FACTS_OPT_SHOW_UNIX_GROUP) {
-    snprintf(ptr, bufsz - buflen, "UNIX.group=%s;",
+    len = snprintf(ptr, bufsz - buflen, "UNIX.group=%s;",
       pr_gid2str(NULL, info->st.st_gid));
-    buflen = strlen(buf);
+    buflen += len;
     ptr = buf + buflen;
   }
 
   if (facts_opts & FACTS_OPT_SHOW_UNIX_MODE) {
-    snprintf(ptr, bufsz - buflen, "UNIX.mode=0%o;",
+    len = snprintf(ptr, bufsz - buflen, "UNIX.mode=0%o;",
       (unsigned int) info->st.st_mode & 07777);
-    buflen = strlen(buf);
+    buflen += len;
     ptr = buf + buflen;
   }
 
   if (facts_opts & FACTS_OPT_SHOW_UNIX_OWNER) {
-    snprintf(ptr, bufsz - buflen, "UNIX.owner=%s;",
+    len = snprintf(ptr, bufsz - buflen, "UNIX.owner=%s;",
       pr_uid2str(NULL, info->st.st_uid));
-    buflen = strlen(buf);
+    buflen += len;
     ptr = buf + buflen;
   }
 
-  /* MLST entries are not sent via pr_data_xfer(), and thus we do not need
-   * to include an LF at the end; it is appended by pr_response_send_raw().
-   * But MLSD entries DO need the trailing LF, so that it can be converted
-   * into a CRLF sequence by pr_data_xfer().
-   */
-  if (strcmp(session.curr_cmd, C_MLSD) == 0) {
-    snprintf(ptr, bufsz - buflen, " %s\n", info->path);
+  if (flags & FACTS_MLINFO_FL_APPEND_CRLF) {
+    len = snprintf(ptr, bufsz - buflen, " %s\r\n", info->path);
```

Comment:
For future reviewers: this change here is the key.  Notice how the previous code path was only using LF, instead of the necessary CRLF.  The previous code path worked, because it relied on code in pr_data_xfer() to convert that LF into a CRLF.  But later changes toggled **off** the ASCII flag, prior to calling pr_data_xfer(), which meant that pr_data_xfer() would **not** perform ASCII conversion.  And that is the cause of this regression.